sql/schemachanger: add dml injection tests for vector indexes

fqazi opened the pull request:
Previously, we didn't have DML injection tests for creating vector
indexes or changing the primary key on tables with vector indexes. We
now add these.

This patch also addresses one more issue in the secondary index recreation logic that can lead to selecting the wrong target index ID, when index chains are being folded.


Informs: #144443
Release notes: none

fqazi committed:
sql/schemachanger: fix RecreateTargetIndexID when deflating indexes

Previously, the logic to rewrite RecreateTargetIndexID could pick the
incorrect index when reducing to fewer indexes. To address, this patch
selects the next suitable index to publish the secondary index with.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/154990)
<!-- Reviewable:end -->

spilchen reviewed (commented):
Looks good. I had one question/comment and a few nits that will help me understand this better.

> @spilchen reviewed 1 of 36 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154990#-:-Ob9KtTQ4eCqtTTNIPgr:b-eyan8y)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/helpers.go` line 1474 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154990#-Ob9GcEjE2Uh8S5eg7TB:-Ob9GcEjE2Uh8S5eg7TC:b1badlu) ([raw file](https://github.com/cockroachdb/cockroach/blob/8817f22071fcdf74320de69ae5a1eaa5b58b6571/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/helpers.go#L1474)):*
> ```Go
> 	old catid.IndexID,
> 	new catid.IndexID,
> 	remaining catid.IndexID,
> ```

nit: I find this parameter name a bit vague. Would something like `newRecreateTarget` be a better name?
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/helpers.go` line 1690 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154990#-Ob9HAho6ID356NILu6q:-Ob9HAho6ID356NILu6r:b-ouclkw) ([raw file](https://github.com/cockroachdb/cockroach/blob/8817f22071fcdf74320de69ae5a1eaa5b58b6571/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/helpers.go#L1690)):*
> ```Go
> 		// as long as its not the old primary index (i.e. we folded all earlier
> 		// primary indexes).
> 		nextIndexID := redundant.SourceIndexID()
> ```

nit: similar comment about the variable name. I suggest a better name. Maybe recreateTargetID?
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/helpers.go` line 1691 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154990#-Ob9ED5-DKQ0NsKLmFCr:-Ob9ED5-DKQ0NsKLmFCs:b-45oqng) ([raw file](https://github.com/cockroachdb/cockroach/blob/8817f22071fcdf74320de69ae5a1eaa5b58b6571/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/helpers.go#L1691)):*
> ```Go
> 		// primary indexes).
> 		nextIndexID := redundant.SourceIndexID()
> 		if nextIndexID == pic.oldSpec.primary.IndexID {
> ```

Shouldn't we also check if nextIndexID is one of the redundants? Like is it possible to have a chain like this:
- inter1 is redundant (being dropped)
- inter2 is redundant (being dropped)
- inter2.SourceIndexID() returns inter1.IndexID

If inter2.SourceIndex() doesn't match the old primary, then we wouldn't pick a new `nextIndexID`. But shouldn't it? inter1 is going to be dropped.


<!-- Sent from Reviewable.io -->

fqazi committed:
sql/schemachanger: add dml injection tests for vector indexes

Previously, we didn't have DML injection tests for creating vector
indexes or changing the primary key on tables with vector indexes. We
now add these.

Informs: #144443
Release notes: none

fqazi reviewed (commented):
> @fqazi reviewed 1 of 36 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154990)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/helpers.go` line 1691 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154990#-Ob9ED5-DKQ0NsKLmFCr:-ObDvnRv4cvNg6jua2ES:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8817f22071fcdf74320de69ae5a1eaa5b58b6571/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/helpers.go#L1691)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Shouldn't we also check if nextIndexID is one of the redundants? Like is it possible to have a chain like this:
- inter1 is redundant (being dropped)
- inter2 is redundant (being dropped)
- inter2.SourceIndexID() returns inter1.IndexID

If inter2.SourceIndex() doesn't match the old primary, then we wouldn't pick a new `nextIndexID`. But shouldn't it? inter1 is going to be dropped.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

spilchen reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @spilchen reviewed 1 of 33 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154990#-:-ObEUnbLBnzEqsjThKS1:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->
